Projects/BookJSON2/main.go

```diff
@@ -20,89 +21,119 @@ type Author struct {
 	Lastname  string
 }
 
-type Books struct {
+type BookHandler struct {
 	Books []Book `json:"book"`
 }
 
 //fake db
 var books []Book
 
-func (b *Books) book(w http.ResponseWriter, r *http.Request) {
-	// if r.Method == POST
+func (b *BookHandler) book(w http.ResponseWriter, r *http.Request) {
 	switch r.Method {
-	case "POST": //Retriving a book with Id number
+	case http.MethodGet: //Retriving a book with Id number
 		{
-			w.Header().Set("content-type", "application/json")
-			jsonreqestfromweb := &Book{}
-			err := json.NewDecoder(r.Body).Decode(&jsonreqestfromweb)
-			if err != nil {
-				fmt.Println(err)
+
+			val := r.FormValue("id")
+			if val == "" {
+				if err := json.NewEncoder(w).Encode(b.Books); err != nil {
+					log.Println(err)
+					return
+				}
 				return
 			}
 
-			json.NewEncoder(w).Encode(b.Books[jsonreqestfromweb.Id])
-			break
+			if val == "break" {
+				w.WriteHeader(http.StatusBadRequest)
+				return
+			}
+
+			id, err := strconv.Atoi(val)
+			if err != nil {
+				panic(err)
+			}
+
+			defer r.Body.Close()
+			for i, value := range b.Books {
+				if value.Id == id {
+					if err := json.NewEncoder(w).Encode(b.Books[i]); err != nil {
```

Comment:
when for loops contain the index and the iterated object at the left side of the range operator, it creates a copy of the object so that it is safe to be used concurrently.

For that reason it is better to use `value` (which contains a copy of the iterated book) than b.Books[i] (which might run into issues if some other requests are modifying the books array at the same time)
```suggestion
			for _, value := range b.Books {
				if value.Id == id {
					if err := json.NewEncoder(w).Encode(value); err != nil {
```

Reply:
Resolved with commit https://github.com/AmanSarraf/Go-Project/pull/9/commits/e5cc27e532ea86a404ba2b08ce15b5656cbe9ae2